File: rdf-infographic-skill/scripts/html_assembler.py
"""Assemble RDF data into a self-contained HTML infographic."""

from __future__ import annotations
import json
import os
import re
import subprocess
import sys
from pathlib import Path
from html import escape
from urllib.parse import quote

try:
    from jinja2 import Environment, FileSystemLoader, Template
    HAS_JINJA = True
except ImportError:
    HAS_JINJA = False
    from string import Template as StrTemplate

from rdf_parser import build_kgdata, extract_narrative, get_base_iri, validate_orphans


HERE = Path(__file__).parent
TEMPLATES_DIR = HERE / "templates"
VALIDATOR = HERE / "validate-harness-contract.py"


def load_asset(name: str) -> str:
    path = TEMPLATES_DIR / name
    if not path.exists():
        raise FileNotFoundError(f"Missing template asset: {path}")
    return path.read_text(encoding="utf-8")


def make_resolver_link(iri: str, resolver_pattern: str = "https://linkeddata.uriburner.com/describe/?url=") -> str:
    return resolver_pattern + quote(iri, safe="")


def make_section_html(section_id: str, title: str, inner_html: str) -> str:
    return (
        f'<section class="section section-alt" id="{section_id}">'
        f'<h2>{title}<a class="headline-anchor" href="#{section_id}" aria-label="Link to this section">¶</a></h2>'
        f'{inner_html}'
        f'</section>'
    )


def render_narrative(rdf_path: str | Path, base_iri: str, resolver_pattern: str) -> tuple[str, list[dict]]:
    """Extract and render narrative sections from RDF annotations."""
    narrative = extract_narrative(rdf_path, base_iri)
    nav_links = [
        {"href": "#hero", "label": "Overview"},
    ]
    html_parts = []
    sections = []

    has_faq = len(narrative["faq"]) > 0
    has_glossary = len(narrative["glossary"]) > 0
    has_howto = len(narrative["howto"]) > 0
    has_people = len(narrative["people"]) > 0
    has_orgs = len(narrative["organizations"]) > 0

    if has_people:
        items_html = ""
        for p in narrative["people"]:
            iri = p["iri"]
            name = escape(p["name"])
            desc = escape(p["description"]) if p["description"] else ""
            items_html += (
                f'<div class="card">'
                f'<h3><a href="{make_resolver_link(iri, resolver_pattern)}" target="_blank" rel="noopener noreferrer">{name}</a></h3>'
                f'<p>{desc}</p></div>'
            )
        html_parts.append(render_narrative_section("people", "People", f'<div class="cards-grid">{items_html}</div>'))
        nav_links.append({"href": "#people", "label": "People"})
        sections.append("people")

    if has_orgs:
        items_html = ""
        for o in narrative["organizations"]:
            iri = o["iri"]
            name = escape(o["name"])
            desc = escape(o["description"]) if o["description"] else ""
            items_html += (
                f'<div class="card">'
                f'<h3><a href="{make_resolver_link(iri, resolver_pattern)}" target="_blank" rel="noopener noreferrer">{name}</a></h3>'
                f'<p>{desc}</p></div>'
            )
        html_parts.append(render_narrative_section("organizations", "Organizations", f'<div class="cards-grid">{items_html}</div>'))
        nav_links.append({"href": "#organizations", "label": "Organizations"})
        sections.append("organizations")

    if has_faq:
        items_html = '<div class="faq-list">'
        for faq in narrative["faq"]:
            iri = faq["iri"]
            q = escape(faq["question"])
            a = escape(faq["answer"])
            link_open = f'<a href="{make_resolver_link(iri, resolver_pattern)}" target="_blank" rel="noopener noreferrer">' if iri else ""
            link_close = "</a>" if iri else ""
            items_html += (
                f'<div class="faq-item anim-fade">'
                f'<div class="faq-question">{link_open}{q}{link_close}<span class="faq-chevron">▼</span></div>'
                f'<div class="faq-answer"><p>{a}</p></div>'
                f'</div>'
            )
        items_html += "</div>"
        html_parts.append(render_narrative_section("faq", "Frequently Asked Questions", items_html))
        nav_links.append({"href": "#faq", "label": "FAQ"})
        sections.append("faq")

    if has_glossary:
        items_html = '<div class="glossary-grid">'
        for g in narrative["glossary"]:
            iri = g["iri"]
            term = escape(g["term"])
            defn = escape(g["definition"])
            link_open = f'<a href="{make_resolver_link(iri, resolver_pattern)}" target="_blank" rel="noopener noreferrer">' if iri else ""
            link_close = "</a>" if iri else ""
            items_html += (
                f'<div class="glossary-term">'
                f'<h4>{link_open}{term}{link_close}</h4>'
                f'<p>{defn}</p></div>'
            )
        items_html += "</div>"
        html_parts.append(render_narrative_section("glossary", "Glossary of Terms", items_html))
        nav_links.append({"href": "#glossary", "label": "Glossary"})
        sections.append("glossary")

    if has_howto:
        items_html = '<div class="howto-list">'
        for i, step in enumerate(narrative["howto"], 1):
            iri = step["iri"]
            s = escape(step["step"])
            desc = escape(step["description"]) if step["description"] else ""
            link_open = f'<a href="{make_resolver_link(iri, resolver_pattern)}" target="_blank" rel="noopener noreferrer">' if iri else ""
            link_close = "</a>" if iri else ""
            items_html += (
                f'<div class="howto-step anim-fade">'
                f'<div class="howto-num">{i}</div>'
                f'<div class="howto-content">'
                f'<h4>{link_open}{s}{link_close}</h4>'
                f'<p>{desc}</p></div></div>'
            )
        items_html += "</div>"
        html_parts.append(render_narrative_section("howto", "How-To Guide", items_html))
        nav_links.append({"href": "#howto", "label": "HowTo"})
        sections.append("howto")

    nav_links.extend([
        {"href": "#kg-explorer", "label": "KG Explorer"},
        {"href": "#sparql-explorer", "label": "SPARQL"},
        {"href": "#footer", "label": "Footer"},
    ])

    return "\n".join(html_parts), nav_links, sections


def render_narrative_section(section_id: str, title: str, inner_html: str) -> str:
    return make_section_html(section_id, title, inner_html)


def render_jsonld(title: str, description: str, base_iri: str, rdf_rel_path: str) -> str:
    ld = {
        "@context": {
            "@vocab": "http://schema.org/",
            "@language": "en",
        },
        "@type": "Article",
        "@id": base_iri,
        "headline": title,
        "description": description,
        "mainEntity": {
            "@type": "CreativeWork",
            "@id": base_iri,
        },
        "sameAs": rdf_rel_path,
    }
    return json.dumps(ld, indent=2)


def build_sparql_recipes(base_iri: str) -> list[dict]:
    return [
        {
            "label": "All triples (sample)",
            "query": f"SELECT ?s ?p ?o\nWHERE {{ ?s ?p ?o }}\nLIMIT 25",
        },
        {
            "label": "Entity types summary",
            "query": f"SELECT ?type (COUNT(?s) AS ?count)\nWHERE {{ ?s a ?type }}\nGROUP BY ?type\nORDER BY DESC(?count)",
        },
        {
            "label": "Named graph triples",
            "query": f"SELECT ?s ?p ?o\nFROM <{base_iri}>\nWHERE {{ ?s ?p ?o }}\nLIMIT 25",
        },
    ]


def assemble_html(
    rdf_path: str | Path,
    output_path: str | Path,
    title: str = "",
    description: str = "",
    source_url: str = "",
    source_label: str = "",
    resolver_pattern: str = "https://linkeddata.uriburner.com/describe/?url=",
    tagline: str = "",
    hero_tagline: str = "",
    meta_html: str = "",
) -> bool:
    """Assemble a complete HTML infographic from an RDF file.

    Returns True on success, False on failure.
    """
    rdf_path = Path(rdf_path)
    output_path = Path(output_path)
    stem = rdf_path.stem

    # Resolve base IRI
    base_iri = get_base_iri(rdf_path)

    if not title:
        title = f"Knowledge Graph Infographic — {stem}"
    if not description:
        description = f"Interactive infographic generated from {rdf_path.name}"

    # Compute relative path for RDF link from output
    rdf_rel = os.path.relpath(str(rdf_path.resolve()), start=str(output_path.parent.resolve()))
    rdf_filename = rdf_path.name

    # Build kgData
    print("Parsing RDF...")
    kgdata = build_kgdata(rdf_path)
    print(f"  Nodes: {len(kgdata['nodes'])}, Links: {len(kgdata['links'])}")

    # Validate orphans
    orphans = validate_orphans(kgdata)
    if orphans:
        print(f"  Warning: {len(orphans)} orphan nodes — {orphans}")
    else:
        print("  Zero orphan nodes")

    # Render narrative
    print("Extracting narrative...")
    narrative_html, nav_links, sections = render_narrative(rdf_path, base_iri, resolver_pattern)
    print(f"  Sections: {', '.join(sections)}")

    # Build JSON-LD
    jsonld_content = render_jsonld(title, description, base_iri, rdf_rel)

    # Build SPARQL recipes
    sparql_recipes = build_sparql_recipes(base_iri)
    default_sparql = sparql_recipes[0]["query"]

    # Load assets
    css_content = load_asset("styles.css")
    kg_explorer_js = load_asset("kg_explorer.js")

    # Serialize kgData
    kgdata_json = json.dumps(kgdata, separators=(",", ":"))

    # Template context
    context = {
        "title": title,
        "description": description,
        "tagline": tagline,
        "hero_tagline": hero_tagline,
        "meta_html": meta_html,
        "rdf_rel_path": rdf_rel,
        "rdf_filename": rdf_filename,
        "output_filename": output_path.name,
        "base_iri": base_iri,
        "css_content": css_content,
        "jsonld_content": jsonld_content,
        "kgdata_json": kgdata_json,
        "kg_explorer_js": kg_explorer_js,
        "nav_links": nav_links,
        "narrative_html": narrative_html,
        "sparql_recipes": sparql_recipes,
        "default_sparql": default_sparql,
        "source_url": source_url,
        "source_label": source_label,
    }

    # Render template
    print("Assembling HTML...")
    if HAS_JINJA:
        env = Environment(
            loader=FileSystemLoader(str(TEMPLATES_DIR)),
            autoescape=False,
        )
        template = env.get_template("base_template.html")
        html = template.render(**context)
    else:
        # Fallback to string.Template
        template_str = load_asset("base_template.html")
        # Convert Jinja2 syntax to $var syntax
        template_str = re.sub(r"\{\{ (\w+) \}\}", r"$\1", template_str)
        template_str = re.sub(r"\{% for (\w+) in (\w+) %\}(.*?)\{% endfor %\}", r"<!-- loop: \1 in \2 -->\3<!-- end loop -->", template_str, flags=re.S)
        template_str = re.sub(r"\{% if (.*?) %\}(.*?)\{% endif %\}", r"\2", template_str, flags=re.S)
        template = StrTemplate(template_str)
        html = template.safe_substitute(**{k: str(v) for k, v in context.items()})

    # Write output
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(html, encoding="utf-8")
    print(f"Written: {output_path} ({output_path.stat().st_size / 1024:.1f} KB)")

    # Validate
    if VALIDATOR.exists():
        print("Running harness contract validation...")
        cmd = [sys.executable, str(VALIDATOR), str(output_path), "--ttl", str(rdf_path)]
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode == 0:
            print("  PASS")
            return True
        else:
            print("  FAIL")
            print("  " + result.stdout.replace("\n", "\n  "))
            return False
    else:
        print(f"  Validator not found at {VALIDATOR}")
        return True

File: rdf-infographic-skill/scripts/html_assembler.ts
/**
 * HTML Assembler — TypeScript edition (Node.js ≥ 18, requires n3).
 * Assembles RDF data into a self-contained HTML infographic.
 * Mirrors html_assembler.py — same signature, same output.
 *
 * Install: npm install (from rdf-infographic-skill/scripts/)
 * Import:  import { assembleHtml } from "./html_assembler.ts";
 */

import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { basename, dirname, join, relative, resolve } from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";
import {
  buildKgdata,
  extractNarrative,
  getBaseIri,
  validateOrphans,
} from "./rdf_parser.ts";

const __dirname_compat = dirname(
  typeof __filename !== "undefined"
    ? __filename
    : fileURLToPath(import.meta.url),
);
const TEMPLATES_DIR = join(__dirname_compat, "templates");
const VALIDATOR    = join(__dirname_compat, "validate-harness-contract.ts");

// ── Mini Jinja2-compatible template engine ────────────────────────────────────

function escHtml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function resolvePath(ctx: Record<string, unknown>, path: string): unknown {
  const parts = path.split(".");
  let val: unknown = ctx;
  for (const p of parts) {
    if (val == null || typeof val !== "object") return "";
    val = (val as Record<string, unknown>)[p];
  }
  return val ?? "";
}

function renderTemplate(tpl: string, ctx: Record<string, unknown>): string {
  // 1. {% for item in list %}...{% endfor %}
  tpl = tpl.replace(
    /\{%-?\s*for\s+(\w+)\s+in\s+(\w+)\s*-?%\}([\s\S]*?)\{%-?\s*endfor\s*-?%\}/g,
    (_, item, listName, body) => {
      const list = resolvePath(ctx, listName);
      if (!Array.isArray(list)) return "";
      return list
        .map((el: unknown) => renderTemplate(body, { ...ctx, [item]: el }))
        .join("");
    },
  );

  // 2. {% if var %}...{% endif %} (truthy check; no else branch needed)
  tpl = tpl.replace(
    /\{%-?\s*if\s+([\w.]+)\s*-?%\}([\s\S]*?)\{%-?\s*endif\s*-?%\}/g,
    (_, cond, body) =>
      resolvePath(ctx, cond) ? renderTemplate(body, ctx) : "",
  );

  // 3. {{ a or b }} — Jinja2-style fallback expression
  tpl = tpl.replace(
    /\{\{\s*([\w.]+)\s+or\s+([\w.]+)\s*\}\}/g,
    (_, a, b) => String(resolvePath(ctx, a) || resolvePath(ctx, b) || ""),
  );

  // 4. {{ var|e }} — HTML-escaped output
  tpl = tpl.replace(
    /\{\{\s*([\w.]+)\|e\s*\}\}/g,
    (_, path) => escHtml(String(resolvePath(ctx, path))),
  );

  // 5. {{ var }} and {{ obj.prop }} — raw output
  tpl = tpl.replace(
    /\{\{\s*([\w.]+)\s*\}\}/g,
    (_, path) => String(resolvePath(ctx, path)),
  );

  return tpl;
}

// ── Template asset loader ─────────────────────────────────────────────────────

function loadAsset(name: string): string {
  const assetPath = join(TEMPLATES_DIR, name);
  if (!existsSync(assetPath)) throw new Error(`Missing template asset: ${assetPath}`);
  return readFileSync(assetPath, "utf-8");
}

// ── Narrative builder ─────────────────────────────────────────────────────────

interface NavLink { href: string; label: string; }
interface SparqlRecipe { label: string; query: string; }

function makeSectionHtml(id: string, title: string, inner: string): string {
  return (
    `<section class="section section-alt" id="${id}">` +
    `<h2>${title}<a class="headline-anchor" href="#${id}" aria-label="Link to this section">¶</a></h2>` +
    `${inner}` +
    `</section>`
  );
}

function resolverLink(iri: string, pattern: string): string {
  return pattern + encodeURIComponent(iri);
}

function renderNarrative(
  rdfPath: string,
  baseIri: string,
  resolverPattern: string,
): { html: string; navLinks: NavLink[]; sections: string[] } {
  const narrative = extractNarrative(rdfPath, baseIri);
  const navLinks: NavLink[] = [{ href: "#hero", label: "Overview" }];
  const htmlParts: string[] = [];
  const sections: string[] = [];

  if (narrative.people.length) {
    let inner = "";
    for (const p of narrative.people) {
      const href = resolverLink(p.iri, resolverPattern);
      inner += `<div class="card"><h3><a href="${href}" target="_blank" rel="noopener noreferrer">${escHtml(p.name)}</a></h3><p>${escHtml(p.description)}</p></div>`;
    }
    htmlParts.push(makeSectionHtml("people", "People", `<div class="cards-grid">${inner}</div>`));
    navLinks.push({ href: "#people", label: "People" });
    sections.push("people");
  }

  if (narrative.organizations.length) {
    let inner = "";
    for (const o of narrative.organizations) {
      const href = resolverLink(o.iri, resolverPattern);
      inner += `<div class="card"><h3><a href="${href}" target="_blank" rel="noopener noreferrer">${escHtml(o.name)}</a></h3><p>${escHtml(o.description)}</p></div>`;
    }
    htmlParts.push(makeSectionHtml("organizations", "Organizations", `<div class="cards-grid">${inner}</div>`));
    navLinks.push({ href: "#organizations", label: "Organizations" });
    sections.push("organizations");
  }

  if (narrative.faq.length) {
    let inner = '<div class="faq-list">';
    for (const faq of narrative.faq) {
      const lo = faq.iri ? `<a href="${resolverLink(faq.iri, resolverPattern)}" target="_blank" rel="noopener noreferrer">` : "";
      const lc = faq.iri ? "</a>" : "";
      inner +=
        `<div class="faq-item anim-fade">` +
        `<div class="faq-question">${lo}${escHtml(faq.question)}${lc}<span class="faq-chevron">▼</span></div>` +
        `<div class="faq-answer"><p>${escHtml(faq.answer)}</p></div>` +
        `</div>`;
    }
    inner += "</div>";
    htmlParts.push(makeSectionHtml("faq", "Frequently Asked Questions", inner));
    navLinks.push({ href: "#faq", label: "FAQ" });
    sections.push("faq");
  }

  if (narrative.glossary.length) {
    let inner = '<div class="glossary-grid">';
    for (const g of narrative.glossary) {
      const lo = g.iri ? `<a href="${resolverLink(g.iri, resolverPattern)}" target="_blank" rel="noopener noreferrer">` : "";
      const lc = g.iri ? "</a>" : "";
      inner += `<div class="glossary-term"><h4>${lo}${escHtml(g.term)}${lc}</h4><p>${escHtml(g.definition)}</p></div>`;
    }
    inner += "</div>";
    htmlParts.push(makeSectionHtml("glossary", "Glossary of Terms", inner));
    navLinks.push({ href: "#glossary", label: "Glossary" });
    sections.push("glossary");
  }

  if (narrative.howto.length) {
    let inner = '<div class="howto-list">';
    narrative.howto.forEach((step, i) => {
      const lo = step.iri ? `<a href="${resolverLink(step.iri, resolverPattern)}" target="_blank" rel="noopener noreferrer">` : "";
      const lc = step.iri ? "</a>" : "";
      inner +=
        `<div class="howto-step anim-fade">` +
        `<div class="howto-num">${i + 1}</div>` +
        `<div class="howto-content"><h4>${lo}${escHtml(step.step)}${lc}</h4><p>${escHtml(step.description)}</p></div>` +
        `</div>`;
    });
    inner += "</div>";
    htmlParts.push(makeSectionHtml("howto", "How-To Guide", inner));
    navLinks.push({ href: "#howto", label: "HowTo" });
    sections.push("howto");
  }

  navLinks.push(
    { href: "#kg-explorer", label: "KG Explorer" },
    { href: "#sparql-explorer", label: "SPARQL" },
    { href: "#footer", label: "Footer" },
  );

  return { html: htmlParts.join("\n"), navLinks, sections };
}

function renderJsonLd(
  title: string,
  description: string,
  baseIri: string,
  rdfRelPath: string,
): string {
  return JSON.stringify(
    {
      "@context": { "@vocab": "http://schema.org/", "@language": "en" },
      "@type": "Article",
      "@id": baseIri,
      headline: title,
      description,
      mainEntity: { "@type": "CreativeWork", "@id": baseIri },
      sameAs: rdfRelPath,
    },
    null,
    2,
  );
}

function buildSparqlRecipes(baseIri: string): SparqlRecipe[] {
  return [
    {
      label: "All triples (sample)",
      query: "SELECT ?s ?p ?o\nWHERE { ?s ?p ?o }\nLIMIT 25",
    },
    {
      label: "Entity types summary",
      query:
        "SELECT ?type (COUNT(?s) AS ?count)\nWHERE { ?s a ?type }\nGROUP BY ?type\nORDER BY DESC(?count)",
    },
    {
      label: "Named graph triples",
      query: `SELECT ?s ?p ?o\nFROM <${baseIri}>\nWHERE { ?s ?p ?o }\nLIMIT 25`,
    },
  ];
}

// ── Public API ────────────────────────────────────────────────────────────────

export interface AssembleHtmlOptions {
  rdfPath: string;
  outputPath: string;
  title?: string;
  description?: string;
  sourceUrl?: string;
  sourceLabel?: string;
  resolverPattern?: string;
  tagline?: string;
  heroTagline?: string;
  metaHtml?: string;
}

export function assembleHtml(opts: AssembleHtmlOptions): boolean {
  const {
    rdfPath,
    outputPath,
    title: titleOpt = "",
    description: descOpt = "",
    sourceUrl = "",
    sourceLabel = "",
    resolverPattern = "https://linkeddata.uriburner.com/describe/?url=",
    tagline = "",
    heroTagline = "",
    metaHtml = "",
  } = opts;

  const stem   = basename(rdfPath).replace(/\.[^.]+$/, "");
  const title  = titleOpt  || `Knowledge Graph Infographic — ${stem}`;
  const description = descOpt || `Interactive infographic generated from ${basename(rdfPath)}`;

  const baseIri       = getBaseIri(rdfPath);
  const rdfRelPath    = relative(dirname(resolve(outputPath)), resolve(rdfPath));
  const rdfFilename   = basename(rdfPath);
  const outputFilename = basename(outputPath);

  console.log("Parsing RDF...");
  const kgdata  = buildKgdata(rdfPath);
  console.log(`  Nodes: ${kgdata.nodes.length}, Links: ${kgdata.links.length}`);

  const orphans = validateOrphans(kgdata);
  if (orphans.length) {
    console.log(`  Warning: ${orphans.length} orphan nodes — ${orphans.slice(0, 5).join(", ")}`);
  } else {
    console.log("  Zero orphan nodes");
  }

  console.log("Extracting narrative...");
  const { html: narrativeHtml, navLinks, sections } = renderNarrative(
    rdfPath, baseIri, resolverPattern,
  );
  console.log(`  Sections: ${sections.join(", ") || "(none)"}`);

  const jsonldContent  = renderJsonLd(title, description, baseIri, rdfRelPath);
  const sparqlRecipes  = buildSparqlRecipes(baseIri);
  const defaultSparql  = sparqlRecipes[0].query;
  const cssContent     = loadAsset("styles.css");
  const kgExplorerJs   = loadAsset("kg_explorer.js");
  const kgdataJson     = JSON.stringify(kgdata);

  console.log("Assembling HTML...");
  const templateStr = loadAsset("base_template.html");
  const html = renderTemplate(templateStr, {
    title,
    description,
    tagline,
    hero_tagline:    heroTagline,
    meta_html:       metaHtml,
    rdf_rel_path:    rdfRelPath,
    rdf_filename:    rdfFilename,
    output_filename: outputFilename,
    base_iri:        baseIri,
    css_content:     cssContent,
    jsonld_content:  jsonldContent,
    kgdata_json:     kgdataJson,
    kg_explorer_js:  kgExplorerJs,
    nav_links:       navLinks,
    narrative_html:  narrativeHtml,
    sparql_recipes:  sparqlRecipes,
    default_sparql:  defaultSparql,
    source_url:      sourceUrl,
    source_label:    sourceLabel,
  });

  mkdirSync(dirname(resolve(outputPath)), { recursive: true });
  writeFileSync(outputPath, html, "utf-8");
  const sizeKb = (statSync(outputPath).size / 1024).toFixed(1);
  console.log(`Written: ${outputPath} (${sizeKb} KB)`);

  if (existsSync(VALIDATOR)) {
    console.log("Running harness contract validation...");
    const result = spawnSync(
      "npx",
      ["tsx", VALIDATOR, outputPath, "--ttl", rdfPath],
      { encoding: "utf-8", shell: true },
    );
    if (result.status === 0) {
      console.log("  PASS");
      return true;
    }
    console.log("  FAIL");
    if (result.stdout) console.log("  " + result.stdout.replace(/\n/g, "\n  "));
    return false;
  }

  console.log(`  Validator not found at ${VALIDATOR}`);
  return true;
}
